### PR TITLE
Cleanup/1886 cleanup global space in prim tests, replace pull_msg with expect_throw_msg

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

This PR advances #1886 with the following:
- cleaned up /prim tests files of global using statements
- remove the unnecesarry pull_msg functions in the gp_* tests. They can be replaced with EXPECT_THROW_MSG.
- puts functions with non-unique names file `f`, `fun`, `f1` in namespaces (The namespace names are based on the file name)
- removed global variables like `const char* function_ = "function";` from the global namespace of tests
- moved util functions like correct_type_vector in the utils header file
- added missing header guards to test utility files (ode mock observers and hmm test utilities)

No release notes as this is part of an ongoing project that is being done over multiple PRs. Will put release notes in the final one.

## Tests

/

## Side Effects

/

## Checklist

- [x] Math issue #1886 

- [x] Copyright holder: Rok Češnovar, Univ. of Ljubljana

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: test/unit/util.hpp [hunk 1]
```diff
@@ -176,4 +177,25 @@ void expect_same_type() {
 }
 }  // namespace test
 
+namespace stan {
+namespace test {
+template <int R, int C>
+void correct_type_vector(const Eigen::Matrix<double, R, C>& x) {
```
Comment (t4c1) [inline test/unit/util.hpp:183]: [optional] I think a better name would be expect_type_vector.

File: test/unit/math/prim/fun/gp_exponential_cov_test.cpp [hunk 2]
```diff
@@ -196,65 +170,67 @@ TEST(MathPrimMat, domain_err_training_sig_l_gp_exp_cov) {
...
+                   std::domain_error, " magnitude");
+  EXPECT_THROW_MSG(stan::math::gp_exponential_cov(x_2, sigma_bad, l_vec),
+                   std::domain_error, " magnitude");
+
+  EXPECT_THROW_MSG(stan::math::gp_exponential_cov(x_2, x_2, sigma, l_bad),
+                   std::domain_error, " length scale");
+  EXPECT_THROW_MSG(stan::math::gp_exponential_cov(x_2, x_2, sigma_bad, l),
+                   std::domain_error, " magnitude");
+  EXPECT_THROW_MSG(stan::math::gp_exponential_cov(x_2, x_2, sigma_bad, l_bad),
+                   std::domain_error, " magnitude");
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply test/unit/math/prim/fun/gp_exponential_cov_test.cpp:202]: Good call, yeah. Will change just to expect throw.

File: test/unit/math/prim/fun/gp_exponential_cov_test.cpp [hunk 3]
```diff
@@ -282,20 +255,12 @@ TEST(MathPrimMat, nan_error_training_sig_l_gp_exp_cov) {
...
-               std::domain_error);
-  EXPECT_THROW(stan::math::gp_exponential_cov(x, sigma_bad, l),
-               std::domain_error);
-  EXPECT_THROW(stan::math::gp_exponential_cov(x, sigma_bad, l_bad),
-               std::domain_error);
+  EXPECT_THROW_MSG(stan::math::gp_exponential_cov(x, sigma, l_bad),
+                   std::domain_error, " length scale");
+  EXPECT_THROW_MSG(stan::math::gp_exponential_cov(x, sigma_bad, l),
+                   std::domain_error, " magnitude");
+  EXPECT_THROW_MSG(stan::math::gp_exponential_cov(x, sigma_bad, l_bad),
```
Comment (t4c1) [inline test/unit/math/prim/fun/gp_exponential_cov_test.cpp:262]: one more

File: test/unit/math/prim/fun/gp_periodic_cov_test.cpp [hunk 4]
```diff
@@ -814,39 +728,20 @@ TEST(MathPrimMat, nan_domain_error_gp_periodic_cov2) {
...
+  EXPECT_THROW_MSG(stan::math::gp_periodic_cov(x1, x2, sigma_bad, l, p),
+                   std::domain_error, " signal standard deviation");
+  EXPECT_THROW_MSG(stan::math::gp_periodic_cov(x1, x2, sigma, l_bad, p_bad),
+                   std::domain_error, " length-scale");
+  EXPECT_THROW_MSG(stan::math::gp_periodic_cov(x1, x2, sigma_bad, l, p_bad),
+                   std::domain_error, " signal standard deviation");
+  EXPECT_THROW_MSG(stan::math::gp_periodic_cov(x1, x2, sigma_bad, l_bad, p),
+                   std::domain_error, " signal standard deviation");
+  EXPECT_THROW_MSG(stan::math::gp_periodic_cov(x1, x2, sigma_bad, l_bad, p_bad),
+                   std::domain_error, " signal standard deviation");
```
Comment (t4c1) [inline test/unit/math/prim/fun/gp_periodic_cov_test.cpp:744]: more

File: test/unit/math/prim/fun/cov_exp_quad_test.cpp [hunk 5]
```diff
@@ -368,18 +342,12 @@ TEST(MathPrimMat, domain_error_training_sig_l_cov_exp_quad) {
...
-
-  EXPECT_THROW(stan::math::cov_exp_quad(x, sigma, l_bad), std::domain_error);
-  EXPECT_THROW(stan::math::cov_exp_quad(x, sigma_bad, l), std::domain_error);
-  EXPECT_THROW(stan::math::cov_exp_quad(x, sigma_bad, l_bad),
-               std::domain_error);
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x, sigma, l_bad), std::domain_error,
+                   " length scale");
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x, sigma_bad, l), std::domain_error,
+                   " magnitude");
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x, sigma_bad, l_bad),
```
Comment (t4c1) [inline test/unit/math/prim/fun/cov_exp_quad_test.cpp:349]: one more

File: test/unit/math/prim/fun/cov_exp_quad_test.cpp [hunk 6]
```diff
@@ -425,18 +393,13 @@ TEST(MathPrimMat, nan_error_training_sig_l_cov_exp_quad) {
...
-  msg2 = pull_msg(x, sigma_bad, l);
-  msg3 = pull_msg(x, sigma_bad, l_bad);
-  EXPECT_TRUE(std::string::npos != msg1.find(" length scale")) << msg1;
-  EXPECT_TRUE(std::string::npos != msg2.find(" magnitude")) << msg2;
-  EXPECT_TRUE(std::string::npos != msg3.find(" magnitude")) << msg3;
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x, sigma, l_bad), std::domain_error,
+                   " length scale");
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x, sigma_bad, l), std::domain_error,
+                   " magnitude");
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x, sigma_bad, l_bad),
```
Comment (t4c1) [inline test/unit/math/prim/fun/cov_exp_quad_test.cpp:400]: one more

File: test/unit/math/prim/fun/cov_exp_quad_test.cpp [hunk 7]
```diff
@@ -488,20 +451,12 @@ TEST(MathPrimMat, domain_error_cov_exp_quad2) {
...
-               std::domain_error);
-  EXPECT_THROW(stan::math::cov_exp_quad(x1, x2, sigma_bad, l),
-               std::domain_error);
-  EXPECT_THROW(stan::math::cov_exp_quad(x1, x2, sigma_bad, l_bad),
-               std::domain_error);
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x1, x2, sigma, l_bad),
+                   std::domain_error, " length scale");
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x1, x2, sigma_bad, l),
+                   std::domain_error, " magnitude");
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x1, x2, sigma_bad, l_bad),
```
Comment (t4c1) [inline test/unit/math/prim/fun/cov_exp_quad_test.cpp:458]: one more

File: test/unit/math/prim/fun/cov_exp_quad_test.cpp [hunk 8]
```diff
@@ -574,13 +529,12 @@ TEST(MathPrimMat, nan_domain_error_cov_exp_quad2) {
...
-  msg2 = pull_msg(x1, x2, sigma_bad, l);
-  msg3 = pull_msg(x1, x2, sigma_bad, l_bad);
-  EXPECT_TRUE(std::string::npos != msg1.find(" length scale")) << msg1;
-  EXPECT_TRUE(std::string::npos != msg2.find(" magnitude")) << msg2;
-  EXPECT_TRUE(std::string::npos != msg3.find(" magnitude")) << msg3;
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x1, x2, sigma, l_bad),
+                   std::domain_error, " length scale");
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x1, x2, sigma_bad, l),
+                   std::domain_error, " magnitude");
+  EXPECT_THROW_MSG(stan::math::cov_exp_quad(x1, x2, sigma_bad, l_bad),
```
Comment (t4c1) [inline test/unit/math/prim/fun/cov_exp_quad_test.cpp:536]: one more

**Comment by rok-cesnovar** [comment]
Thanks. Done.

**Pull request merged**